### PR TITLE
Added api for downloading the script

### PR DESCRIPTION
**Pull request by vviveksharma**
## Description

- This is api helps us to download the install script according to the persona mentioned in the baseurl and the os_type on which the script is going to run
- API CONTRACT : https://chefio.atlassian.net/wiki/spaces/LA/pages/3013181515/Get+API+for+Downloading+install+scripts
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] If `Gemfile.lock` has changed, I have used `--conservative` to do it and included the full output in the Description above.
- [ ] All new and existing tests passed.
- [ ] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

## DEMOS: 
- API Testing: https://progresssoftware.sharepoint.com/:v:/s/ChefCoreC/EV-P7p0umwBOlj39y1_6WmABWZ4a4OF7ZPJMQiDWTijrqA?e=PfbPhl
- Linux Script Testing : https://progresssoftware.sharepoint.com/:v:/s/ChefCoreC/EfOmKfFLu8NMjCtnztonj7sBO0DIrvaO9aq8U8CNCHrQ9g?e=aKkM0L
- Windows Script Testing: https://progresssoftware.sharepoint.com/:v:/s/ChefCoreC/ETpv1RJj1wdHjOKiAnD8KA8BvOogOgViP4dj-SD8YoathA?nav=eyJyZWZlcnJhbEluZm8iOnsicmVmZXJyYWxBcHAiOiJTdHJlYW1XZWJBcHAiLCJyZWZlcnJhbFZpZXciOiJTaGFyZURpYWxvZy1MaW5rIiwicmVmZXJyYWxBcHBQbGF0Zm9ybSI6IldlYiIsInJlZmVycmFsTW9kZSI6InZpZXcifX0%3D&e=pZSnl8

### CODE REVIEW

File: models/models.go [hunk 1]
```diff
@@ -25,3 +25,8 @@ type RelatedProducts struct {
 	Bom      string            `json:"bom"`
 	Products map[string]string `json:"products"`
 }
+
+type Script struct {
```
Comment (RajeshPaul38) [inline models/models.go:29]: The name can be more appropriate:
```suggestion
type ScriptInput struct {
```
Or
```suggestion
type ScriptParams struct {
```

File: services/main.go [hunk 2]
```diff
@@ -648,3 +649,36 @@ func (server *ApiService) isOsVersion(params *omnitruck.RequestParams, c *fiber.
...
+	err, ok := server.ValidateRequest(params, c)
+	if !ok {
+		server.logCtx(c).Error("Validation of download script API for " + params.Product + " failed")
+		return err
+	}
+	if params.OsType == "" {
+		return server.SendErrorResponse(c, http.StatusBadRequest, "query os_type can't be empty please use linux or windows depending on your operating system.")
+	}
+	if params.OsType != "linux" && params.OsType != "windows" {
+		return server.SendErrorResponse(c, http.StatusBadRequest, "query os_type can only be used as linux or windows depending on your operating system.")
```
Comment (RajeshPaul38) [inline services/main.go:665]: we should use `StatusUnprocessableEntity` as the request structure is not bad it's only the value which we don't support.

File: services/main_test.go [hunk 3]
```diff
@@ -723,3 +725,82 @@ func TestFileNameHandler(t *testing.T) {
...
+			serverMode: 2,
+			mockfileutils: func(baseUrl string, params *omnitruck.RequestParams, filepath string) (string, error) {
+				return "", nil
+			},
+			requestPath:      `/stable/chef/downloadScript?license_id=afd2c0a2-111f-4caf-1fa2-1211fe1212d1&os_type=`,
+			expectedStatus:   400,
+			expectedResponse: "",
+		},
+		{
+			name:       "if os_type is inavalid",
```
Comment (RajeshPaul38) [inline services/main_test.go:760]:
```suggestion
			name:       "if os_type is invalid",
```

File: models/models.go (hunk 1 above)
Comment (vviveksharma) [reply models/models.go:29]: Done

File: services/main_test.go (hunk 3 above)
Comment (vviveksharma) [reply services/main_test.go:760]: Done

File: services/main.go (hunk 2 above)
Comment (vviveksharma) [reply services/main.go]: Done

File: utils/fileutils/fileutils.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+		LicenseId: params.LicenseId,
+	}
+	templateReader, err := template.ParseFiles(filePath)
+	if err != nil {
+		return "", fiber.NewError(http.StatusInternalServerError, "error while parsing the template files: "+err.Error())
+	}
+	var scriptResponse bytes.Buffer
+	err = templateReader.Execute(&scriptResponse, scriptInput)
+	if err != nil {
+		return "", fiber.NewError(http.StatusInternalServerError, "error while executing the tempalte reader object: "+err.Error())
```
Comment (ayush-billore) [inline utils/fileutils/fileutils.go:34]:
```suggestion
		return "", fiber.NewError(http.StatusInternalServerError, "error while executing the template reader object: "+err.Error())
```

**Review by ayush-billore — APPROVED** [review]

File: utils/fileutils/fileutils.go (hunk 4 above)
Comment (vviveksharma) [reply utils/fileutils/fileutils.go]: Done

**Review by vipin230 — APPROVED** [review]

**Review by RoyShravani — APPROVED** [review]

**Review by rishabhjhs — APPROVED** [review]

**Review by sandhi18 — APPROVED** [review]

**Pull request merged**